Audit preparation (#121)
NestedFi/nested-core-lego: README.MD

<div align="center"><img width="400" src="https://user-images.githubusercontent.com/32484870/169264834-35b0831e-c6fd-417e-9675-477424488c60.jpeg"></div>
<br/>

<div align="center"><a href="https://nested.fi" > https://nested.fi </a></div>

# Introduction

Nested is a decentralized protocol providing customizable financial products in the form of NFTs.
The platform allows users to put several digital assets, i.e. ERC20 tokens, inside an NFT (abbreviated as `NestedNFT`).
<br/>

Each NestedNFT is backed by underlying assets:

-   Purchased or sold on a decentralized exchange (AMM).
-   Collected/earned after adding liquidity or staking.
-   Exchanged/Minted on a protocol that is not a decentralized exchange.

> -   (...)

Reviewer: I think it was automatically added by `prettier` (same for other comments on dashes)

> -   (...)

Author: Ok, should we keep these changes @maximebrugel @adrien-supizet ?

> -   (...)

Reviewer 2: no strong opinion, whatever looks best with markdown

| MixinOperatorResolver          | Abstract contract to load authorized operators in cache.                                                                 |
| ZeroExOperator                 | Performs token swaps through 0x ([read more](contracts/operators/ZeroEx/README.md)).                                     |
| ZeroExStorage                  | ZeroExOperator storage contract. Must store the 0x `swapTarget`.                                                         |
| FlatOperator                   | Handles deposits and withdraws. No interaction with any third parties ([read more](contracts/operators/Flat/README.md)). |
| BeefyVaultOperator             | Handles deposits and withdraws in a Beefy single vault (native or non-native).                                           |
| BeefyZapBiswapLPVaultOperator  | Handles deposits and withdraws in a Beefy Biswap LP vault using zapper                                                   |
| BeefyZapUniswapLPVaultOperator | Handles deposits and withdraws in a Beefy UniswapV2 LP vault using zapper                                                |

> | BeefyVaultStorage              | Handles whitelisting of Beefy Vault or Zapper                                                                            |

Reviewer: it misses `.` at the end of the 3 lines above

}
```

> It help us to make **one** interaction, but we want to make multiple interactions. For example, to create a portfolio with multiple tokens, we need to "batch" these orders.

Reviewer 2: ```suggestion
It helps us to make **one** interaction, but we want to make multiple interactions. For example, to create a portfolio with multiple tokens, we need to "batch" these orders.
```

#### Batched Output Orders

<div align="center"><img src="./static/output-orders.png" width="650"></div>

-   Multiple inputs for every orders but one output.

> -   0.3% fee on the output if operation does not reduce TVL, 0.8% if it does

Reviewer: it misses  a `.` at the end

Hence, each time the Nested Factory needs to interact with user funds (which are represented as a NestedNFT), it will first check the balance of tokens associated with the NestedNFT through Nested Records. If needed, it will then transfert funds to the Nested Reserve or withdraw funds from it.

## ETH managment

The Nested protocol only handles WETH when calling operators.  
The sent ETH used to feed a portfolio is automatically converted to WETH when received.

### WETH Conversion

The conversion from ETH to WETH is done when submitting an order through `_submitInOrders` or `_submitOutOrders`.

Before submitting orders, the NestedFactory transfers the input tokens from the NestedReserve (or the msg.sender) to the factory and converts the sent ETH to WETH.

### Use of ETH in operators

There are some operators who use ETH directly and not WETH.  
In this case, the operator uses the Withdrawer to get the ETH back from the WETH contract before using it.

> Only the Withdrawer can send ETH to the NestedFactory without automatic WETH conversion.

## Royalties

> Royalties are withdrawal fees when a user has copied an existing portfolio and are managed by the `FeeSplitter.sol`. Royalties are distributed from the `NestedFactory.sol` when a user withdraws funds from his portfolio OR when an order fails, all tokens are returned to the sender and are therefore subject to withdrawal fees.

Reviewer: ```suggestion
Royalties are withdrawal fees allocated to the owner of the original portfolio when a user has copied an existing portfolio. They are managed by the `FeeSplitter.sol`. Royalties are distributed from the `NestedFactory.sol` when a user withdraws funds from his portfolio OR when an order fails, all tokens are returned to the sender and are therefore subject to withdrawal fees.
```

Reviewer 2: Royalties are a part of the fee collected by the protocol.
Thus royalties are collected during every step of the copied portfolio lifecycle (copy, update, deposit, withdraw) not only upon withdrawing funds

### Fees distribution

> For now, protfolio designers and Nested share equally the royalties.

Reviewer: ```suggestion
For now, portfolio designers and Nested share the royalties equally.
```

Reviewer 2: Royalties are a part of the fees given to the portfolio creators ⚠️ 
See from the docs if you need more information: 
```
First scenario: you create a portfolio from scratch
 In this case, all the fees go to Nested Finance Ltd and there are no royalties.
Second scenario: you replicate another portfolio
Fees are equally shared between Nested Finance Ltd. and the initial creator of the Nested portfolio. 
This distribution does not occur only at the first replication of a portfolio, but during its entire lifecycle. 
````
```suggestion
For now, the fees are shared equally between portfolio creators (as royalties) and Nested.
```

This distribution is done in the `feeSplitter.sol` with the ratio between the `royaltiesWeight` and the `shareholders` weights.
Currently, we have `royaltiesWeight = 50` and **one** shareholder (_Nested_) with `weight = 50`.

> **totalWeights** always equals **royaltiesWeight** + all shareholders **weights**

# Ownership & Governance

Some functions of the protocol require admin rights (`onlyOwner` using `Ownable` from [OpenZeppelin](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/access/Ownable.sol)). Same with the [TransparentUpgradeableProxy](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/proxy/transparent/TransparentUpgradeableProxy.sol) which need an **Admin**.

> <img width="1011" alt="image" src="https://user-images.githubusercontent.com/22816913/173034995-432d6e84-e9d2-40cf-acb3-e109a158240f.png">

Reviewer: I would suggest to upload the image to the directory static (name can be reviewed, `assets` maybe ?).
It will allow us to know from where the image is and have an history.  
You can then display the image like just above
`<div align="center"><img src="./static/output-orders.png" width="650"></div>`

The contracts are owned by the `OwnerProxy` which is a [DSProxy](https://github.com/dapphub/ds-proxy) fork without a cache, where only the Timelock can execute the scripts.

The TimelockControllerEmergency is a [TimelockController](https://docs.openzeppelin.com/contracts/4.x/api/governance#TimelockController) fork. 

> It introduces the "Emergency Role" to execute a transaction in an instantaneous way, but only the "Emergency Multisig" has this role, with 5 members and 5 approvals needed (in the case of an urgent fix).

Reviewer: ```suggestion
It introduces the "Emergency Role" to execute a transaction in an instantaneous way. 
Only the "Emergency Multisig" has this role, with 5 members and 5 approvals needed in the case of an urgent fix.
```

> On the other hand, the "Operational Multisig" can schedule/execute transactions with a **6-hours** delay, with 3 members an 2 approvals needed.

Reviewer: ```suggestion
On the other hand, the "Operational Multisig" can schedule/execute transactions with a **6-hours** delay, with 3 members and 2 approvals needed.
```

}
```

#### Example `processInputOrders` flow

<div align="center"><img src="./static/processInputOrders.png" width="1200"></div>

## Nested Factory interaction with the Nested Reserve and Nested Records
The Nested Reserve stores underlying assets of all NestedNFTs. The Nested Records keeps track of which underlying assets are associated with a specific NestedNFT. 
Hence, each time the Nested Factory needs to interact with user funds (which are represented as a NestedNFT), it will first check the balance of tokens associated with the NestedNFT through Nested Records. If needed, it will then transfert funds to the Nested Reserve or withdraw funds from it.

> ## ETH managment

Reviewer 3: ```suggestion
## Native Token Management
```

ETH and WETH are specific to mainnet. Since the protocol is multi-chain, it's better to replace "ETH" by "Native Token" and "WETH" by "Wrapped Native Token".

The Nested protocol only handles WETH when calling operators.  
The sent ETH used to feed a portfolio is automatically converted to WETH when received.

### WETH Conversion

The conversion from ETH to WETH is done when submitting an order through `_submitInOrders` or `_submitOutOrders`.

Before submitting orders, the NestedFactory transfers the input tokens from the NestedReserve (or the msg.sender) to the factory and converts the sent ETH to WETH.

### Use of ETH in operators

There are some operators who use ETH directly and not WETH.  
In this case, the operator uses the Withdrawer to get the ETH back from the WETH contract before using it.

> Only the Withdrawer can send ETH to the NestedFactory without automatic WETH conversion.

## Royalties
Royalties are a part of the fee collected by the protocol and they are collected during every step of a copied portfolio lifecycle (copy, update, deposit, withdraw).

### Fees distribution
For now, the fees are shared equally between portfolio creators (as royalties) and Nested.

> **First scenario:** When a portfolio is created from scratch, all fees go to Nested Finance Ltd and there are no royalties.

Reviewer 3: ```suggestion
**First scenario:** When a portfolio is created from scratch, all fees go to Nested Finance Ltd and there are no royalties.<br/>
```
